docs/proposals/resource-quota-reservations.md

```diff
@@ -0,0 +1,241 @@
+to allow for optimistic locking.  If the quota usage
+was stale, the admission control logic runs again
+to validate the request is still valid with the latest
+quota document.
+
+A quota controller regularly sweeps the system to
+recalculate observed usage.  This is primarily useful
+for monitoring deleted resources and replenishing
+available quota.  If the quota controller sees a delta
+in its recalculated usage relative to the currently
+reported status, it will update the usage.  If its update
+is stale, it will recalculate usage again.
+
+There is a small window of time that could result in a
+race condition where the quota controller reports an
+incorrect usage.
+
+
+In pseudo-code:
+
+**API_SERVER**
+
+```
+Admission control (quota)
+T1. QUOTA = ...
+T2. IF QUOTA.USED + REQUEST >= QUOTA.HARD
+T3.   REJECT
+T4. UPDATE QUOTA.STATUS(QUOTA.USED + REQUEST)
+T5. ADMIT
+
+RESTStorage
+T6. Create Object
+```
+
+**QUOTA_CONTROLLER**
+
+```
+T1. QUOTA := ...
+T2. OBSERVED_USAGE := ... [live queries based on latest data]
+T3. UPDATE QUOTA.STATUS.USED IFF OBSERVED USAGE != QUOTA.STATUS.USED
+```
+
+If the `QUOTA_CONTROLLER.T1` happens after `API_SERVER.T4`,
+and `QUOTA_CONTROLLER.T3` happens before `API_SERVER.T6` completes,
+then when the quota controller recalculates usage, it could have calculated
+usage without observing the outcome of the create.
+
+In practice, this scenario is *extremely rare*, but could happen if
+there was a network hiccup of some kind that blocked creation of
+the object at time `API_SERVER.T6` in etcd without causing the request
+to timeout.  The API server would have also been responding and reading
+from that same etcd to support a full quota recalculation and status
+update at time `QUOTA_CONTROLLER.T3`.
+
+The next controller pass would recalculate usage correctly.
+
+A solution should be put in place to mitigate this risk and a long-term
+plan to handle should be agreed upon.
+
+## Use Cases
+
+* As a `cluster-admin`, I want more reliable quota enforcement.
+
+## Proposed Change
+
+The proposal discusses a short-term fix to mitigate the race window
+and a long-term solution to solving the problem.  It's important to
+note that multi-object commits in etcd v3 may help solve this solution,
+but if we plan to support federated API servers, we could not depend
+on that solution.
+
+### Mitigation plan
+
+The quota admission controller will add an annotation to a quota
+that tracks when it's usage was last incremented.
+
+The quota controller will not recalculate usage for a quota document
+until some variable amount of time has expired since the last increment.
+
+There is a potential delay in replenishing available quota
+if there are a large number of incremental creates.
+
+The quota controller will still do compare-and-swap so if an
+incremental create occurs in the time that it recalculated usage
+it will be told it has stale data and recalculate.
+
+### ResourceQuota reservations
```

Comment:
@deads2k - this should solve the latent list scenario unless I am mistaken.  the replenishment controller would not release the reservation, so off by N scenarios should not happen in HA.